runtime/compiler/runtime/ClassUnloadAssumption.cpp

```diff
@@ -606,6 +612,58 @@ TR_RuntimeAssumptionTable::assumptionCanBeSerialized(TR_RuntimeAssumptionKind ki
    return canBeSerialized;
    }
 
+uint8_t *
+TR_RuntimeAssumptionTable::serialize(J9JITConfig *jitConfig)
+   {
+   OMR::CriticalSection serializeAssumptions(assumptionTableMutex);
+
+   if (TR::Options::getCmdLineOptions()->getVerboseOption(TR_VerbosePerformance))
+      TR_VerboseLog::writeLineLocked(TR_Vlog_PERF, "Serializing Assumptions:");
+
+   uint8_t *bufferStart = (uint8_t *)TR_PersistentMemory::jitPersistentAlloc(getTotalSizeOfAssumptionsToBeSerialized() + sizeof(SerializedDataAtlas));
+   TR_ASSERT_FATAL(bufferStart, "Failed to allocate memory to serialize runtime assumptions\n");
+
+   SerializedDataAtlas *atlas = (SerializedDataAtlas *)bufferStart;
+   memset(atlas, 0, sizeof(SerializedDataAtlas));
+   atlas->_size = getTotalSizeOfAssumptionsToBeSerialized();
+
+   uint8_t *buffer = bufferStart;
+   buffer += sizeof(SerializedDataAtlas);
+
+   for (int k=0; k < LastAssumptionKind; k++) // for each table
+      {
+      TR_RatHT *hashTable = _tables + k;
+      size_t hashTableSize = hashTable->_spineArraySize;
+      for (size_t i = 0; i < hashTableSize; ++i) // for each bucket
+         {
+         for (OMR::RuntimeAssumption *cursor = hashTable->_htSpineArray[i]; cursor; cursor = cursor->getNext()) // for each entry
+            {
+            if (!cursor->isMarkedForDetach())
+               {
+               TR::SentinelRuntimeAssumption *sentinel = (TR::SentinelRuntimeAssumption *)cursor->getSentinel();
+               TR_ASSERT_FATAL(sentinel, "Sentinel can't be NULL\n");
+
+               uint32_t size = cursor->size();
+               J9JITExceptionTable *owningMetadata = (J9JITExceptionTable *)sentinel->getOwningMetadata();
+               TR_ASSERT_FATAL(owningMetadata, "owningMetadata can't be NULL\n");
+
+               cursor->serialize(buffer, (uint8_t *)owningMetadata);
+
+               atlas->_numAssumptionsPerKind[k]._count++;
+               atlas->_numAssumptionsPerKind[k]._size += size;
+
+               owningMetadata->runtimeAssumptionList = NULL;
+               buffer += size;
+               }
+            }
+         }
+
+      if (!assumptionCanBeSerialized((TR_RuntimeAssumptionKind)k))
+         TR_ASSERT_FATAL(atlas->_numAssumptionsPerKind[k]._count == 0, "Assumption(s) of kind %d serialized!", k);
```

Comment:
I was staring at this for a while wondering how this makes sense given that we're going through all the assumptions and calling serialize on them. Then I remembered that in the `snapshot` branch, when running in snapshot mode, we restrict the JIT in a way that results only in the assumptions we expect to be serialized to get created.

This check is, I guess, a way of validating that the compiler isn't generating assumptions we're not expecting to serialize.